src/Npgsql/NpgsqlConnector.cs

```diff
@@ -607,32 +607,25 @@ async Task RawOpen(NpgsqlTimeout timeout, bool async, CancellationToken cancella
                     case 'S':
                         var clientCertificates = new X509CertificateCollection();
                         var certPath = Settings.ClientCertificate ?? PostgresEnvironment.SslCert;
-                        var certPathExists = true;
 
-                        if (certPath is null)
-                        {
-                            certPath = PostgresEnvironment.SslCertDefault;
-                            certPathExists = File.Exists(certPath);
-                        }
+                        if (certPath is null && PostgresEnvironment.SslCertDefault is { } certPathDefault && File.Exists(certPathDefault))
+                            certPath = certPathDefault;
 
-                        if (certPathExists)
-                            clientCertificates.Add(new X509Certificate(certPath!));
+                        if (certPath != null)
+                            clientCertificates.Add(new X509Certificate2(certPath, Settings.ClientCertificateKey ?? PostgresEnvironment.SslKey));
```

Comment:
BTW it seems that X509Certificate2 is disposable... we should make sure to dispose it.

Reply:
`X509Certificate` is disposable too, but internally they use safe handles, so it's safe not to dispose them. At the end the used certificate should be disposed by the stream.

Reply:
I don't think we should depend on such internal implementation details (which could change). Any reason not to dispose?

Reply:
Should we in that case dispose user provided certificates too?

Reply:
I don't think so, because they're user-provided, and so they're owned by the user (for example, they could be using the same certificate across many connection instances). But for any disposable objects that we allocate ourselves, we should dispose them.

Reply:
BTW should we invoke `ProvideClientCertificatesCallback` if there is a certificate already?

Reply:
Good point. It's possible to pass multiple certificate to SslStream, right? Any idea what it was, does PG check if *any* of them is works? If so, then always invoking ProvideClientCertificatesCallback makes sense to me. Otherwise we need to decide on an ordering a bit like with the password/pgpass stuff...

Reply:
It uses only one provided file.

Reply:
Any idea why clientCertificate is a collection there?

BTW I'm really just asking questions to understand - I don't know much about SslStream and the certificates side of things. If it's really only one certificate file that gets checked by PostgreSQL, then we need to apply precedence rules - probably the exact same logic we just applied for passwords.